File: 0039-combination-sum/0039-combination-sum.cpp
class Solution {
private:
    void findCombination(int index, vector<int> candidates, vector<int> &temp, vector<vector<int>> &result, int target){
        if(target < 0 || index == candidates.size()) return ;

        if(target == 0){
            result.push_back(temp);
            return;
        }

        for(int i=index;i<candidates.size(); i++){
            temp.push_back(candidates[i]);
            findCombination(i,candidates,temp,result,target-candidates[i]);
            temp.pop_back();
        }
    }
public:
    vector<vector<int>> combinationSum(vector<int>& candidates, int target) {
        vector<int> temp;
        vector<vector<int>> result;

        findCombination(0,candidates,temp,result,target);
        return result;
    }
};

File: 0039-combination-sum/0039-combination-sum.java
class Solution {
    private void findCombination(int index, int[] candidates, int target,List<List<Integer>> result, List<Integer> temp){
        if(index == candidates.length){
            if(target == 0){
                result.add(new ArrayList<>(temp));
            }
            return;
        }

        if(target-candidates[index] >=0){
            temp.add(candidates[index]); // pick
            findCombination(index,candidates,target-candidates[index],result,temp);
            temp.remove(temp.size()-1); // backtract
        }
        findCombination(index+1, candidates, target, result, temp); // non pick with the next index
    }
    public List<List<Integer>> combinationSum(int[] candidates, int target) {
        List<List<Integer>> result = new ArrayList<>();
        List<Integer> temp = new ArrayList<>();
        findCombination(0,candidates,target,result,temp);
        return result;
    }
}